### PR TITLE
Implement array indexes and field references in Expressions 

### PR DESCRIPTION
**Pull request by aswan**

### CODE REVIEW

File: expr/expr.go [hunk 1]
```diff
@@ -801,3 +823,93 @@ func compileArithmetic(lhsFunc, rhsFunc NativeEvaluator, operator string) (Nativ
...
+		rhs, err := rhsFunc(rec)
+		if err != nil {
+			return NativeValue{}, err
+		}
+
+		var idx uint
+		switch rhs.typ.ID() {
+		case zng.IdByte, zng.IdUint16, zng.IdUint32, zng.IdUint64:
+			idx = uint(rhs.value.(uint64))
+		case zng.IdInt16, zng.IdInt32, zng.IdInt64:
```
Comment (nwt) [inline expr/expr.go]: Any reason not to do `switch rhs.type {`?

File: zql/parser-support.go [hunk 2]
```diff
@@ -281,6 +281,15 @@ func makeGroupByProc(durationIn, limitIn, keysIn, reducersIn interface{}) *ast.G
 	}
 }
 
+// XXX explain me
```
Comment (nwt) [inline zql/parser-support.go]:
```suggestion
```

File: expr/expr.go [hunk 3]
```diff
@@ -801,3 +823,93 @@ func compileArithmetic(lhsFunc, rhsFunc NativeEvaluator, operator string) (Nativ
 		}
 	}, nil
 }
+
+func getNthFromContainer(container zcode.Bytes, idx uint) (zcode.Bytes, error) {
```
[the remark this answers is not in the document]
Comment (aswan) [reply expr/expr.go]: Its also conventional (at least outside of go, if go has some different convention I think it should be ignored here) to use unsigned when a negative value isn't meaningful, such as retrieving the nth item from an ordered list...

File: expr/expr.go (hunk 1 above)
Comment (aswan) [reply expr/expr.go]: Not sure what you're suggesting, `rhs.type` isn't a thing.  You mean `rhs.typ.(type)` ?  The immediate problem with that is that it doesn't resolve aliases (there was a lengthy discussion about this on a different PR)

File: expr/expr.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (nwt) [reply expr/expr.go]: It's conventional in Go. The standard library, for instance, usually prefers int to uint for parameters and return values that cannot be negative.

[the remark this answers is not in the document]
Comment (aswan) [reply expr/expr.go]: Well again, I don't think we gain anything from following this convention here.
If you really want to follow it though, can you educate me on how the convention works?  Do I then add an explicit `if idx < 0` test as the first line?

[the remark this answers is not in the document]
Comment (nwt) [reply expr/expr.go]: > Well again, I don't think we gain anything from following this convention here.

This was just a nit, so feel free to disregard it.

> If you really want to follow it though, can you educate me on how the convention works? Do I then add an explicit if idx < 0 test as the first line?

I'd just do `s/idx uint/idx int/` and `s/var i uint = 0/var i int/`.

File: expr/expr.go (hunk 1 above)
Comment (nwt) [reply expr/expr.go]: I mean `rhs.typ`. And I'd forgotten about aliases. 🤦‍♂

**Pull request merged**